File: scraper.py
#!/usr/bin/env python3
"""
TOEIC文法問題スクレイパー - zitanstudy.com
Usage: python scraper.py
Output: questions.json

サイト構造:
  トップ  : https://zitanstudy.com/?page_id=206
  インデックス: https://zitanstudy.com/?page_id=4190  (第1〜100問)
  個別問題: https://zitanstudy.com/?page_id=210      (第1問)
"""

import json
import time
import re
import logging
import sys
from pathlib import Path
from urllib.parse import urljoin

import requests
from bs4 import BeautifulSoup

# ── ログ設定 ───────────────────────────────────────────────────────────────────
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)-8s %(message)s",
    handlers=[
        logging.StreamHandler(sys.stdout),
        logging.FileHandler("scraper.log", encoding="utf-8"),
    ],
)
log = logging.getLogger(__name__)

# ── 定数 ──────────────────────────────────────────────────────────────────────
BASE        = "https://zitanstudy.com"
TOP_PAGE    = f"{BASE}/?page_id=206"
FIRST_INDEX = f"{BASE}/?page_id=4190"
OUTPUT      = Path("questions.json")
PROGRESS    = Path(".scraper_progress.json")
DELAY       = 1.5  # リクエスト間隔（秒）

HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/120.0.0.0 Safari/537.36"
    ),
    "Accept": "text/html,application/xhtml+xml;q=0.9,*/*;q=0.8",
    "Accept-Language": "ja,en-US;q=0.9,en;q=0.8",
    "Referer": BASE + "/",
    "Connection": "keep-alive",
}

sess = requests.Session()
sess.headers.update(HEADERS)


# ── HTTPフェッチ ──────────────────────────────────────────────────────────────
def fetch(url: str, tries: int = 3) -> BeautifulSoup | None:
    for i in range(tries):
        try:
            r = sess.get(url, timeout=20)
            r.raise_for_status()
            r.encoding = r.apparent_encoding or "utf-8"
            return BeautifulSoup(r.text, "html.parser")
        except requests.HTTPError as e:
            code = e.response.status_code
            log.warning(f"HTTP {code}: {url}")
            if code in (403, 404):
                return None
        except Exception as e:
            log.warning(f"エラー ({url}): {e}")
        time.sleep(2 ** (i + 1))
    log.error(f"フェッチ失敗: {url}")
    return None


# ── コンテンツ要素取得 ─────────────────────────────────────────────────────────
def main_content(soup: BeautifulSoup):
    """記事本文要素を返す。見つからなければ body を返す。"""
    for sel in [
        "div.entry-content",
        "div.post-content",
        "div.article-body",
        "article",
        "main",
    ]:
        el = soup.select_one(sel)
        if el:
            return el
    return soup.body or soup


# ── インデックスページURL収集 ──────────────────────────────────────────────────
def collect_index_urls(soup: BeautifulSoup) -> list[str]:
    """トップページからインデックスページURLを収集する。"""
    urls, seen = [], set()
    for a in main_content(soup).find_all("a", href=True):
        href = a["href"]
        if "page_id=" not in href:
            continue
        if "page_id=206" in href:  # トップページ自身を除外
            continue
        u = urljoin(BASE, href)
        text = a.get_text(strip=True)
        if u not in seen and re.search(r'\d', text):
            urls.append(u)
            seen.add(u)
    return urls


# ── 問題URLを収集 ──────────────────────────────────────────────────────────────
def collect_question_urls(soup: BeautifulSoup) -> list[str]:
    """インデックスページから個別問題ページURLを収集する。"""
    urls, seen = [], set()
    for a in main_content(soup).find_all("a", href=True):
        href = a["href"]
        if "page_id=" not in href:
            continue
        u = urljoin(BASE, href)
        text = a.get_text(strip=True)
        # 「第N問」「N問」「数字のみ」のリンクを問題リンクと判定
        if u not in seen and re.search(r'(第?\s*\d+\s*問|\b\d+\b)', text):
            urls.append(u)
            seen.add(u)
    return urls


# ── 問題ページパース ───────────────────────────────────────────────────────────
def parse_question(soup: BeautifulSoup, url: str, fallback_id: int) -> dict | None:
    """問題ページを解析して辞書形式で返す。"""
    root = main_content(soup)

    # テキストブロックを収集（子要素を持つコンテナは除外して重複を防ぐ）
    blocks = []
    for el in root.descendants:
        if el.name not in ("p", "h1", "h2", "h3", "h4", "h5", "li", "td"):
            continue
        if el.find(["p", "h1", "h2", "h3", "h4", "li"]):
            continue
        t = el.get_text(" ", strip=True)
        if t and len(t) > 2:
            blocks.append(t)

    if not blocks:
        log.warning(f"コンテンツなし: {url}")
        return None

    # 正規表現パターン
    P_NUM  = re.compile(r'第\s*(\d+)\s*問')
    P_CHOI = re.compile(r'^[\s　]*[\(（]([ABCDabcd])[\)）][.\s　]*(.+)')
    P_ANS  = re.compile(r'答[えい][\s　]*[：:：][\s　]*[\(（]?([ABCDabcd])[\)）]?')
    P_TRAN = re.compile(r'^(訳|和訳|日本語訳|【訳】|翻訳)[\s：:：]*')
    P_EXPL = re.compile(r'^(解説|【解説】|ポイント)[\s：:：]*')
    P_VOCE = re.compile(r'^(語彙|単語|【語彙】|【単語】|Words?)[\s：:：]*')
    P_NAV  = re.compile(r'^(HOME|TOEIC|Copyright|サイト|メニュー|Menu|Skip)', re.I)

    qid  = fallback_id
    qtxt = ""
    choices = {"A": "", "B": "", "C": "", "D": ""}
    ans   = ""
    trans = ""
    expl_parts: list[str] = []
    vocab_parts: list[str] = []
    state = "pre"

    for blk in blocks:
        # 問題番号
        m = P_NUM.search(blk)
        if m and state == "pre":
            qid = int(m.group(1))
            if re.fullmatch(r'[^\w]*第\s*\d+\s*問[^\w]*', blk.strip()):
                state = "q"
                continue

        # 選択肢
        m = P_CHOI.match(blk)
        if m:
            choices[m.group(1).upper()] = m.group(2).strip()
            state = "choices"
            continue

        # 答え
        m = P_ANS.search(blk)
        if m:
            ans = m.group(1).upper()
            state = "ans"
            continue

        # 訳
        if P_TRAN.match(blk):
            trans = P_TRAN.sub("", blk).strip()
            state = "trans"
            continue

        # 解説
        if P_EXPL.match(blk):
            body = P_EXPL.sub("", blk).strip()
            if body:
                expl_parts.append(body)
            state = "expl"
            continue

        # 語彙
        if P_VOCE.match(blk):
            body = P_VOCE.sub("", blk).strip()
            if body:
                vocab_parts.append(body)
            state = "vocab"
            continue

        # 状態別の蓄積
        if state in ("pre", "q"):
            if len(blk) > 10 and not P_NAV.match(blk):
                qtxt = (qtxt + " " + blk).strip()
                state = "q"
        elif state == "trans":
            trans = (trans + " " + blk).strip()
        elif state == "expl":
            expl_parts.append(blk)
        elif state == "vocab":
            vocab_parts.append(blk)

    if not qtxt and not any(choices.values()):
        log.warning(f"Q{qid}: コンテンツ抽出失敗 {url}")
        return None

    return {
        "id":          qid,
        "question":    qtxt.strip(),
        "choices":     choices,
        "answer":      ans,
        "translation": trans.strip(),
        "explanation": "\n".join(expl_parts).strip(),
        "vocabulary":  "\n".join(vocab_parts).strip(),
        "source_url":  url,
    }


# ── 進捗の保存・読み込み ────────────────────────────────────────────────────────
def load_progress() -> tuple[set, list]:
    if PROGRESS.exists():
        d = json.loads(PROGRESS.read_text("utf-8"))
        return set(d.get("done", [])), d.get("qs", [])
    return set(), []


def save_progress(done: set, qs: list) -> None:
    PROGRESS.write_text(
        json.dumps({"done": sorted(done), "qs": qs}, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )


# ── メイン ────────────────────────────────────────────────────────────────────
def main() -> None:
    log.info("=== TOEIC Grammar Scraper (zitanstudy.com) ===")

    done, qs = load_progress()
    log.info(f"進捗読み込み: {len(done)} 完了, {len(qs)} 問保存済み")

    # Step 1: トップページからインデックスURLを収集
    log.info(f"トップページ取得: {TOP_PAGE}")
    top_soup = fetch(TOP_PAGE)
    time.sleep(DELAY)

    idx_urls: list[str] = []
    if top_soup:
        idx_urls = collect_index_urls(top_soup)
        log.info(f"インデックスページ: {len(idx_urls)} 件")

    # 既知の最初のインデックスページを必ず含める
    if FIRST_INDEX not in idx_urls:
        idx_urls.insert(0, FIRST_INDEX)

    if not idx_urls:
        log.error("インデックスページが見つかりません。終了します。")
        sys.exit(1)

    # Step 2: 各インデックスページから問題URLを収集
    q_urls: list[str] = []
    q_seen: set[str]  = set()

    for i_url in idx_urls:
        log.info(f"インデックス取得: {i_url}")
        i_soup = fetch(i_url)
        time.sleep(DELAY)
        if i_soup:
            found = collect_question_urls(i_soup)
            for u in found:
                if u not in q_seen:
                    q_urls.append(u)
                    q_seen.add(u)
            log.info(f"  → {len(found)} 問リンク (累計 {len(q_urls)})")

    log.info(f"スクレイプ対象: {len(q_urls)} 問")

    # Step 3: 各問題ページをスクレイプ
    for i, q_url in enumerate(q_urls, 1):
        if q_url in done:
            log.debug(f"[{i}/{len(q_urls)}] スキップ (取得済み): {q_url}")
            continue

        log.info(f"[{i}/{len(q_urls)}] {q_url}")
        q_soup = fetch(q_url)

        if q_soup:
            q = parse_question(q_soup, q_url, i)
            if q:
                qs.append(q)
                log.info(f"  Q{q['id']}: {q['question'][:60]}…")

        done.add(q_url)

        if i % 10 == 0:
            save_progress(done, qs)
            log.info(f"  中間保存: {len(qs)} 問完了")

        time.sleep(DELAY)

    # 最終保存
    qs.sort(key=lambda x: x["id"])
    OUTPUT.write_text(
        json.dumps(qs, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )
    log.info(f"✓ {len(qs)} 問を {OUTPUT} に保存しました")

    if PROGRESS.exists():
        PROGRESS.unlink()


if __name__ == "__main__":
    main()

File: scraper.js
/**
 * TOEIC Grammar Scraper — ブラウザコンソール版
 *
 * 【使い方】
 * 1. https://zitanstudy.com/?page_id=4190 をブラウザで開く
 * 2. F12 → Console タブ（Macは Cmd+Option+J）
 * 3. このスクリプトをまるごとコピー＆ペーストしてEnter
 * 4. 完了すると questions.json が自動ダウンロードされる
 *
 * ※ スマホ（Android）の場合は Kiwi Browser を使うとコンソールが使えます
 * ※ スマホ（iOS）の場合はブックマークレット版（下部参照）を使ってください
 */

(async () => {
  const BASE  = 'https://zitanstudy.com';
  const DELAY = 700; // ms（サーバー負荷軽減）
  const sleep = ms => new Promise(r => setTimeout(r, ms));
  const p     = new DOMParser();

  const parse  = html => p.parseFromString(html, 'text/html');
  const getDoc = async url => parse(await (await fetch(url)).text());

  function mainContent(doc) {
    return doc.querySelector('.entry-content,.post-content,article,main') || doc.body;
  }

  function textBlocks(doc) {
    return [...mainContent(doc).querySelectorAll('p,h1,h2,h3,h4,li')]
      .filter(el => !el.querySelector('p,h1,h2,h3,h4,li'))
      .map(el => el.textContent.replace(/\s+/g, ' ').trim())
      .filter(t => t.length > 2);
  }

  function parseQuestion(doc, url, fid) {
    const blocks = textBlocks(doc);
    const RE = {
      num:  /第\s*(\d+)\s*問/,
      choi: /^[\s　]*[（(]([ABCDabcd])[）)][.\s　]*(.+)/,
      ans:  /答[えい][\s　]*[：:]\s*[（(]?([ABCDabcd])[）)]/,
      tran: /^(訳|和訳|日本語訳|【訳】)/,
      expl: /^(解説|【解説】)/,
      voce: /^(語彙|単語|【語彙】)/,
      nav:  /^(HOME|TOEIC|Copyright|サイト)/i,
    };

    let qid = fid, qtxt = '', ans = '', trans = '';
    const choices = { A:'', B:'', C:'', D:'' };
    const expl = [], vocab = [];
    let state = 'pre';

    for (const blk of blocks) {
      let m;
      if ((m = RE.num.exec(blk)) && state === 'pre') {
        qid = +m[1];
        if (/^[^\w]*第\s*\d+\s*問[^\w]*$/.test(blk)) { state = 'q'; continue; }
      }
      if ((m = RE.choi.exec(blk))) {
        choices[m[1].toUpperCase()] = m[2].trim(); state = 'choices'; continue;
      }
      if ((m = RE.ans.exec(blk)))  { ans = m[1].toUpperCase(); state = 'ans'; continue; }
      if (RE.tran.test(blk)) { trans = blk.replace(RE.tran,'').replace(/^[\s：:]*/,''); state='trans'; continue; }
      if (RE.expl.test(blk)) { const b=blk.replace(RE.expl,'').replace(/^[\s：:]*/,''); if(b)expl.push(b); state='expl'; continue; }
      if (RE.voce.test(blk)) { const b=blk.replace(RE.voce,'').replace(/^[\s：:]*/,''); if(b)vocab.push(b); state='voce'; continue; }

      if (state==='pre'||state==='q') { if(blk.length>10&&!RE.nav.test(blk)){qtxt=(qtxt+' '+blk).trim();state='q';} }
      else if (state==='trans') { trans+=' '+blk; state='expl'; }
      else if (state==='expl')  expl.push(blk);
      else if (state==='voce')  vocab.push(blk);
    }

    if (!qtxt && !Object.values(choices).some(Boolean)) return null;
    return {
      id: qid, question: qtxt.trim(), choices, answer: ans,
      translation: trans.trim(),
      explanation: expl.join('\n').trim(),
      vocabulary:  vocab.join('\n').trim(),
      source_url:  url,
    };
  }

  // ── Step 1: インデックスページURLを収集 ──────────────────────────────────────
  console.log('📚 TOEIC Scraper 開始...');
  const topDoc  = await getDoc(`${BASE}/?page_id=206`);
  await sleep(DELAY);

  const idxUrls = [...new Set([
    `${BASE}/?page_id=4190`,
    ...[...mainContent(topDoc).querySelectorAll('a[href*="page_id="]')]
      .map(a => a.href)
      .filter(h => !h.includes('page_id=206')),
  ])];
  console.log(`インデックスページ: ${idxUrls.length}件`);

  // ── Step 2: 問題URLを収集 ───────────────────────────────────────────────────
  const qUrls = [], seen = new Set();
  for (const iu of idxUrls) {
    const doc = await getDoc(iu);
    [...mainContent(doc).querySelectorAll('a[href*="page_id="]')]
      .filter(a => /第?\s*\d+\s*問|\b\d+\b/.test(a.textContent))
      .forEach(a => { if (!seen.has(a.href)) { qUrls.push(a.href); seen.add(a.href); } });
    console.log(`  収集中... ${qUrls.length}問`);
    await sleep(DELAY);
  }
  console.log(`問題総数: ${qUrls.length}問`);

  // ── Step 3: 各問題ページをパース ─────────────────────────────────────────────
  const questions = [];
  for (let i = 0; i < qUrls.length; i++) {
    const doc = await getDoc(qUrls[i]);
    const q   = parseQuestion(doc, qUrls[i], i + 1);
    if (q) questions.push(q);
    if ((i + 1) % 20 === 0)
      console.log(`進捗: ${i+1}/${qUrls.length} — ${questions.length}問取得`);
    await sleep(DELAY);
  }

  questions.sort((a, b) => a.id - b.id);

  // ── ダウンロード ──────────────────────────────────────────────────────────────
  const blob = new Blob([JSON.stringify(questions, null, 2)], { type: 'application/json' });
  Object.assign(document.createElement('a'), {
    href: URL.createObjectURL(blob),
    download: 'questions.json',
  }).click();

  console.log(`✅ 完了！ ${questions.length}問を questions.json としてダウンロードしました`);
  console.log('📖 questions.json を index.html と同じフォルダに置いてサイトを開いてください');
})();


/* ═══════════════════════════════════════════════════════════════════════════════
   📱 iPhone（iOS Safari）用ブックマークレット
   ─────────────────────────────────────────────────────────────────────────────
   【セットアップ手順】
   1. Safari で https://zitanstudy.com/?page_id=4190 を開いてブックマーク登録
      （共有ボタン → "ブックマークを追加"）
   2. ブックマーク一覧を開き、今追加したブックマークを「編集」
   3. 名前を「TOEIC Scraper」などに変更
   4. URL 欄を全部消して、↓の javascript: から始まる1行をまるごと貼り付けて保存

   【使い方】
   1. Safari で https://zitanstudy.com/?page_id=4190 を開く
   2. アドレスバーに「TOEIC」と入力してブックマークを選択（タップ）
   3. 「スクレイピング開始」のアラートが出たら OK → 数十分待つ
   4. 完了アラートが出たら画面に JSON テキストが表示される
   5. 「全選択してコピー」ボタンをタップ → コピー完了
   6. index.html を開き「JSONを読み込む」→ 貼り付け → 読み込む
   ═══════════════════════════════════════════════════════════════════════════════

javascript:(async()=>{const B='https://zitanstudy.com',D=700,s=ms=>new Promise(r=>setTimeout(r,ms)),p=new DOMParser(),g=async u=>p.parseFromString(await(await fetch(u)).text(),'text/html'),c=d=>d.querySelector('.entry-content,article,main')||d.body,b=d=>[...c(d).querySelectorAll('p,h1,h2,h3,h4,li')].filter(e=>!e.querySelector('p,h1,h2,h3,h4,li')).map(e=>e.textContent.replace(/\s+/g,' ').trim()).filter(t=>t.length>2);function q(d,u,f){const bl=b(d),R={n:/第\s*(\d+)\s*問/,c:/^[\s　]*[（(]([ABCDabcd])[）)][.\s　]*(.+)/,a:/答[えい][\s　]*[：:]\s*[（(]?([ABCDabcd])[）)]/,t:/^(訳|和訳)/,e:/^(解説)/,v:/^(語彙|単語)/};let id=f,qt='',an='',tr='';const ch={A:'',B:'',C:'',D:''},ex=[],vc=[];let st='pre';for(const k of bl){let m;if((m=R.n.exec(k))&&st==='pre'){id=+m[1];if(/^[^\w]*第\s*\d+\s*問/.test(k)){st='q';continue;}}if((m=R.c.exec(k))){ch[m[1].toUpperCase()]=m[2].trim();st='c';continue;}if((m=R.a.exec(k))){an=m[1].toUpperCase();st='a';continue;}if(R.t.test(k)){tr=k.replace(R.t,'').replace(/^[\s：:]*/,'');st='t';continue;}if(R.e.test(k)){const x=k.replace(R.e,'').replace(/^[\s：:]*/,'');if(x)ex.push(x);st='e';continue;}if(R.v.test(k)){const x=k.replace(R.v,'').replace(/^[\s：:]*/,'');if(x)vc.push(x);st='v';continue;}if(st==='pre'||st==='q'){if(k.length>10&&!/^(HOME|TOEIC|Copyright)/i.test(k)){qt=(qt+' '+k).trim();st='q';}}else if(st==='t'){tr+=' '+k;st='e';}else if(st==='e')ex.push(k);else if(st==='v')vc.push(k);}if(!qt&&!Object.values(ch).some(Boolean))return null;return{id,question:qt.trim(),choices:ch,answer:an,translation:tr.trim(),explanation:ex.join('\n').trim(),vocabulary:vc.join('\n').trim(),source_url:u};}alert('スクレイピング開始！画面はそのまま待ってください（数十分かかります）');const td=await g(`${B}/?page_id=206`);await s(D);const iu=[...new Set([`${B}/?page_id=4190`,...[...c(td).querySelectorAll('a[href*="page_id="]')].map(a=>a.href).filter(h=>!h.includes('page_id=206'))])];const qu=[],se=new Set();for(const i of iu){const d=await g(i);[...c(d).querySelectorAll('a[href*="page_id="]')].filter(a=>/第?\s*\d+\s*問|\b\d+\b/.test(a.textContent)).forEach(a=>{if(!se.has(a.href)){qu.push(a.href);se.add(a.href);}});await s(D);}const qs=[];for(let i=0;i<qu.length;i++){const d=await g(qu[i]),r=q(d,qu[i],i+1);if(r)qs.push(r);await s(D);}qs.sort((a,b)=>a.id-b.id);const json=JSON.stringify(qs);const ov=document.createElement('div');ov.style.cssText='position:fixed;inset:0;background:#fff;z-index:99999;display:flex;flex-direction:column;padding:16px;font-family:sans-serif';ov.innerHTML=`<div style="font-weight:700;font-size:1.1rem;margin-bottom:8px">✅ 完了！${qs.length}問取得</div><p style="font-size:.85rem;color:#555;margin-bottom:10px">①下のボタンで全選択してコピー → ②index.htmlの「JSONを読み込む」に貼り付け</p><button onclick="const t=this.nextElementSibling;t.select();document.execCommand('copy');this.textContent='✅ コピーしました！'" style="background:#2563eb;color:#fff;border:none;padding:10px 20px;border-radius:8px;font-size:1rem;font-weight:700;margin-bottom:10px;cursor:pointer">全選択してコピー</button><textarea readonly style="flex:1;border:1px solid #ccc;border-radius:8px;padding:10px;font-size:.75rem;font-family:monospace">${json}</textarea>`;document.body.appendChild(ov);})();

*/
